python/ray/tune/tests/test_tune_save_restore.py

```diff
@@ -0,0 +1,160 @@
+# coding: utf-8
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import pickle
+import shutil
+import tempfile
+import unittest
+
+import ray
+from ray import tune
+from ray.rllib import _register_all
+from ray.tune import Trainable
+
+
+class SerialTuneRelativeLocalDirTest(unittest.TestCase):
+    local_mode = True
+    prefix = "Serial"
+
+    class MockTrainable(Trainable):
+        _name = "MockTrainable"
+
+        def _setup(self, config):
+            self.state = {"hi": 1}
+
+        def _train(self):
+            return {"timesteps_this_iter": 1, "done": True}
+
+        def _save(self, checkpoint_dir):
+            checkpoint_path = os.path.join(
+                checkpoint_dir, "checkpoint-{}".format(self._iteration))
+            with open(checkpoint_path, "wb") as f:
+                pickle.dump(self.state, f)
+            return checkpoint_path
+
+        def _restore(self, checkpoint_path):
+            with open(checkpoint_path, "rb") as f:
+                extra_data = pickle.load(f)
+            self.state.update(extra_data)
+
+    def setUp(self):
+        ray.init(num_cpus=1, num_gpus=0, local_mode=self.local_mode)
+
+    def tearDown(self):
+        ray.shutdown()
+
+        # Without this line, test_tune_server.testAddTrial would fail.
```

Comment:
What's the error that we get here without `_register_all()`? It sounds like a bug, and we should have to call it.

Reply:
`_register_all()` is so widely used in other tests of tune, and it mostly used at `tearDown()`. I think it's a useful workaround.

Here's the error of `test_tune_server` after removing `_register_all()`:

```
FTrying port 4321
Tearing down....
2019-05-07 01:32:07,147	INFO node.py:497 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-07_01-32-07_146512_78456/logs.
2019-05-07 01:32:07,259	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:16900 to respond...
2019-05-07 01:32:07,376	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:55880 to respond...
2019-05-07 01:32:07,380	INFO services.py:806 -- Starting Redis shard with 3.44 GB max memory.
2019-05-07 01:32:07,397	INFO node.py:511 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-07_01-32-07_146512_78456/logs.
2019-05-07 01:32:07,398	INFO services.py:1441 -- Starting the Plasma object store with 5.15 GB memory using /tmp.
2019-05-07 01:32:07,476	INFO web_server.py:241 -- Starting Tune Server...

ray/tune/tests/test_tune_server.py:61 (TuneServerSuite.testAddTrial)
self = <test_tune_server.TuneServerSuite testMethod=testAddTrial>

    def testAddTrial(self):
>       runner, client = self.basicSetup()

test_tune_server.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_tune_server.py:46: in basicSetup
    trials = [Trial("__fake", **kwargs), Trial("__fake", **kwargs)]
../trial.py:271: in __init__
    Trial._registration_check(trainable_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'ray.tune.trial.Trial'>, trainable_name = '__fake'

    @classmethod
    def _registration_check(cls, trainable_name):
        if not has_trainable(trainable_name):
            # Make sure rllib agents are registered
            from ray import rllib  # noqa: F401
            if not has_trainable(trainable_name):
>               raise TuneError("Unknown trainable: " + trainable_name)
E               ray.tune.error.TuneError: Unknown trainable: __fake

../trial.py:340: TuneError

```

Reply:
I see. @richardliaw do you have any thoughts on this?

Reply:
Yeah, `_register_all` right now allows us to use `__fake`, which is a Mock Trainable. We need to register after we shutdown Ray because we only do registration upon import, which happens once in the beginning of execution (and we shutdown/clear plasma multiple times).

One thing we should do in a separate PR is to move `__fake` (or some version of this) into Tune.

Reply:
Sounds good! Thanks!